{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GRPH"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Given: A collection of DNA strings in FASTA format having total length at most 10 kbp.\n",
    "\n",
    "Return: The adjacency list corresponding to O3. You may return edges in any order."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rosalind_6539 Rosalind_6866\n",
      "Rosalind_6539 Rosalind_7971\n",
      "Rosalind_6539 Rosalind_7682\n",
      "Rosalind_0386 Rosalind_0854\n",
      "Rosalind_0386 Rosalind_0170\n",
      "Rosalind_6477 Rosalind_6381\n",
      "Rosalind_6477 Rosalind_9341\n",
      "Rosalind_6477 Rosalind_1678\n",
      "Rosalind_8636 Rosalind_5784\n",
      "Rosalind_1303 Rosalind_5156\n",
      "Rosalind_4410 Rosalind_5461\n",
      "Rosalind_4410 Rosalind_5552\n",
      "Rosalind_9424 Rosalind_5755\n",
      "Rosalind_9424 Rosalind_4356\n",
      "Rosalind_0584 Rosalind_5949\n",
      "Rosalind_0584 Rosalind_0508\n",
      "Rosalind_2131 Rosalind_4818\n",
      "Rosalind_2131 Rosalind_0704\n",
      "Rosalind_4693 Rosalind_5461\n",
      "Rosalind_4693 Rosalind_5552\n",
      "Rosalind_5247 Rosalind_1439\n",
      "Rosalind_5777 Rosalind_8821\n",
      "Rosalind_5777 Rosalind_5685\n",
      "Rosalind_5755 Rosalind_1970\n",
      "Rosalind_5755 Rosalind_7673\n",
      "Rosalind_5435 Rosalind_6539\n",
      "Rosalind_5435 Rosalind_6949\n",
      "Rosalind_8828 Rosalind_0584\n",
      "Rosalind_8828 Rosalind_4468\n",
      "Rosalind_8828 Rosalind_6302\n",
      "Rosalind_5461 Rosalind_1303\n",
      "Rosalind_5461 Rosalind_9188\n",
      "Rosalind_8807 Rosalind_0584\n",
      "Rosalind_8807 Rosalind_4468\n",
      "Rosalind_8807 Rosalind_6302\n",
      "Rosalind_0041 Rosalind_3168\n",
      "Rosalind_0041 Rosalind_8575\n",
      "Rosalind_0041 Rosalind_2062\n",
      "Rosalind_6381 Rosalind_7431\n",
      "Rosalind_6381 Rosalind_9492\n",
      "Rosalind_7508 Rosalind_6203\n",
      "Rosalind_7508 Rosalind_9472\n",
      "Rosalind_7508 Rosalind_8212\n",
      "Rosalind_7809 Rosalind_1694\n",
      "Rosalind_7809 Rosalind_5249\n",
      "Rosalind_7809 Rosalind_3095\n",
      "Rosalind_1220 Rosalind_6381\n",
      "Rosalind_1220 Rosalind_9341\n",
      "Rosalind_1220 Rosalind_1678\n",
      "Rosalind_5413 Rosalind_0854\n",
      "Rosalind_5413 Rosalind_0170\n",
      "Rosalind_6866 Rosalind_7508\n",
      "Rosalind_6866 Rosalind_9441\n",
      "Rosalind_0854 Rosalind_3168\n",
      "Rosalind_0854 Rosalind_8575\n",
      "Rosalind_0854 Rosalind_2062\n",
      "Rosalind_4356 Rosalind_5461\n",
      "Rosalind_4356 Rosalind_5552\n",
      "Rosalind_2988 Rosalind_9721\n",
      "Rosalind_3168 Rosalind_0386\n",
      "Rosalind_3168 Rosalind_5247\n",
      "Rosalind_7431 Rosalind_2575\n",
      "Rosalind_7431 Rosalind_9110\n",
      "Rosalind_7971 Rosalind_8828\n",
      "Rosalind_7971 Rosalind_9700\n",
      "Rosalind_7971 Rosalind_7794\n",
      "Rosalind_9697 Rosalind_6866\n",
      "Rosalind_9697 Rosalind_7971\n",
      "Rosalind_9697 Rosalind_7682\n",
      "Rosalind_8821 Rosalind_5413\n",
      "Rosalind_8821 Rosalind_2125\n",
      "Rosalind_8570 Rosalind_9721\n",
      "Rosalind_6203 Rosalind_1439\n",
      "Rosalind_9341 Rosalind_3168\n",
      "Rosalind_9341 Rosalind_8575\n",
      "Rosalind_9341 Rosalind_2062\n",
      "Rosalind_2534 Rosalind_7643\n",
      "Rosalind_0170 Rosalind_5413\n",
      "Rosalind_0170 Rosalind_2125\n",
      "Rosalind_4818 Rosalind_9721\n",
      "Rosalind_9188 Rosalind_2131\n",
      "Rosalind_9188 Rosalind_4578\n",
      "Rosalind_9188 Rosalind_7653\n",
      "Rosalind_9188 Rosalind_5429\n",
      "Rosalind_9721 Rosalind_7726\n",
      "Rosalind_8575 Rosalind_4626\n",
      "Rosalind_5552 Rosalind_0386\n",
      "Rosalind_5552 Rosalind_5247\n",
      "Rosalind_4948 Rosalind_8232\n",
      "Rosalind_4948 Rosalind_9697\n",
      "Rosalind_5685 Rosalind_3168\n",
      "Rosalind_5685 Rosalind_8575\n",
      "Rosalind_5685 Rosalind_2062\n",
      "Rosalind_4626 Rosalind_7726\n",
      "Rosalind_4578 Rosalind_0386\n",
      "Rosalind_4578 Rosalind_5247\n",
      "Rosalind_0467 Rosalind_3312\n",
      "Rosalind_9941 Rosalind_4446\n",
      "Rosalind_0263 Rosalind_9972\n",
      "Rosalind_9700 Rosalind_0584\n",
      "Rosalind_9700 Rosalind_4468\n",
      "Rosalind_9700 Rosalind_6302\n",
      "Rosalind_3237 Rosalind_0467\n",
      "Rosalind_3237 Rosalind_5661\n",
      "Rosalind_6949 Rosalind_1694\n",
      "Rosalind_6949 Rosalind_5249\n",
      "Rosalind_6949 Rosalind_3095\n",
      "Rosalind_9492 Rosalind_4948\n",
      "Rosalind_5156 Rosalind_4818\n",
      "Rosalind_5156 Rosalind_0704\n",
      "Rosalind_4352 Rosalind_9721\n",
      "Rosalind_2206 Rosalind_4693\n",
      "Rosalind_2206 Rosalind_5777\n",
      "Rosalind_7653 Rosalind_5949\n",
      "Rosalind_7653 Rosalind_0508\n",
      "Rosalind_5784 Rosalind_1970\n",
      "Rosalind_5784 Rosalind_7673\n",
      "Rosalind_1678 Rosalind_4948\n",
      "Rosalind_5293 Rosalind_8636\n",
      "Rosalind_5293 Rosalind_8807\n",
      "Rosalind_9972 Rosalind_5461\n",
      "Rosalind_9972 Rosalind_5552\n",
      "Rosalind_2363 Rosalind_9931\n",
      "Rosalind_2363 Rosalind_5293\n",
      "Rosalind_5249 Rosalind_0386\n",
      "Rosalind_5249 Rosalind_5247\n",
      "Rosalind_0704 Rosalind_9941\n",
      "Rosalind_0704 Rosalind_2206\n",
      "Rosalind_0704 Rosalind_0795\n",
      "Rosalind_7682 Rosalind_3312\n",
      "Rosalind_2125 Rosalind_0584\n",
      "Rosalind_2125 Rosalind_4468\n",
      "Rosalind_2125 Rosalind_6302\n",
      "Rosalind_5661 Rosalind_4410\n",
      "Rosalind_2575 Rosalind_5949\n",
      "Rosalind_2575 Rosalind_0508\n",
      "Rosalind_9472 Rosalind_8828\n",
      "Rosalind_9472 Rosalind_9700\n",
      "Rosalind_9472 Rosalind_7794\n",
      "Rosalind_3524 Rosalind_0386\n",
      "Rosalind_3524 Rosalind_5247\n",
      "Rosalind_4468 Rosalind_5435\n",
      "Rosalind_0508 Rosalind_2363\n",
      "Rosalind_1308 Rosalind_6203\n",
      "Rosalind_1308 Rosalind_9472\n",
      "Rosalind_1308 Rosalind_8212\n",
      "Rosalind_2062 Rosalind_9931\n",
      "Rosalind_2062 Rosalind_5293\n",
      "Rosalind_0795 Rosalind_4693\n",
      "Rosalind_0795 Rosalind_5777\n",
      "Rosalind_2629 Rosalind_6866\n",
      "Rosalind_2629 Rosalind_7971\n",
      "Rosalind_2629 Rosalind_7682\n",
      "Rosalind_7643 Rosalind_5949\n",
      "Rosalind_7643 Rosalind_0508\n",
      "Rosalind_3312 Rosalind_9931\n",
      "Rosalind_3312 Rosalind_5293\n",
      "Rosalind_6302 Rosalind_2131\n",
      "Rosalind_6302 Rosalind_4578\n",
      "Rosalind_6302 Rosalind_7653\n",
      "Rosalind_6302 Rosalind_5429\n",
      "Rosalind_7794 Rosalind_3312\n",
      "Rosalind_3095 Rosalind_5949\n",
      "Rosalind_3095 Rosalind_0508\n",
      "Rosalind_5429 Rosalind_5461\n",
      "Rosalind_5429 Rosalind_5552\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "dna_strings = \"rosalind_grph.txt\"\n",
    "n = 3\n",
    "\n",
    "seq_name, seq_string = [], []\n",
    "#parse file\n",
    "with open (dna_strings,'r') as fa:\n",
    "    for seq_record  in SeqIO.parse(fa,'fasta'):\n",
    "        seq_name.append(str(seq_record.name))\n",
    "        seq_string.append(str(seq_record.seq))\n",
    "for i in range(len(seq_string)):\n",
    "    for j in range(len(seq_string)):\n",
    "        if i != j:\n",
    "            if seq_string[i][-n:] == seq_string[j][:n]:\n",
    "                print(seq_name[i], seq_name[j])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"./answers/roslaind_lcsm_outFile.txt\", \"w\") as outFile:\n",
    "        print(lcs, file=outFile)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MPRT\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Fix code before submittion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P01374_TNFB_HUMAN\n",
      "96 \n",
      "P01190_COLI_BOVIN\n",
      "91 \n",
      "P0AF66\n",
      "90 121 \n",
      "P03415_VME1_CVMA5\n",
      "27 \n",
      "Q03745\n",
      "272 438 506 509 550 1043 1096 \n",
      "P37803\n",
      "110 \n",
      "P01044_KNH1_BOVIN\n",
      "47 87 168 169 197 204 \n",
      "A9M5H3\n",
      "133 \n"
     ]
    }
   ],
   "source": [
    "import requests as rq\n",
    "text = open(\"rosalind_mprt.txt\", \"rt\")\n",
    "data = text.readlines()\n",
    "text.close()\n",
    "\n",
    "\n",
    "def hasmotifN(s):\n",
    "    l1 =[]\n",
    "    for i in range(len(s)):\n",
    "        if i<len(s)-3: \n",
    "            substr = s[i:i+4]\n",
    "            if substr[0] == 'N':\n",
    "                if substr[1] != 'P':\n",
    "                    if (substr[2] == 'S' or substr[2] == 'T'):\n",
    "                        if substr[3] !='P':\n",
    "                            l1.append(i)\n",
    "    return l1  \n",
    "\n",
    "for i in data:  \n",
    "    URL = \"http://www.uniprot.org/uniprot/\" + str(i.strip('\\n')[0:6]) + \".fasta\"\n",
    "    response = rq.get(URL)\n",
    "    l = str(response.text)\n",
    "    l = l.split('\\n')\n",
    "    l2 = []\n",
    "    for j in l:\n",
    "        if j!='':\n",
    "            if j[0] == '>':\n",
    "                l2.append('')\n",
    "            if j[0] != '>': \n",
    "                l2 += j\n",
    "    s = ''.join(l2)\n",
    "    a = hasmotifN(s)\n",
    "    if a != []:\n",
    "        print(i.strip('\\n'))\n",
    "    for m in a:\n",
    "        print(m+1,end=' ')\n",
    "    if a!= []:\n",
    "        print('\\n',end='')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ORF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MRA\n",
      "MNGDTIQIRLVATSRI\n",
      "MTGCPKDWDLIPRR\n",
      "MLIPRLVLAPN\n",
      "MPRRILSPLHYRKLHIIRSSQG\n",
      "MYMLCPHYPCEERIICSFL\n",
      "MSEDG\n",
      "MATPFRFGS\n",
      "MYTKGYGLKRPVIWPISSGMPCGLKLG\n",
      "MKIHLYSFRNTKLQCTTMTGCPKDWDLIPRR\n",
      "MVSP\n",
      "MDWRPTGTLDAVASCASVLG\n",
      "MPAR\n",
      "M\n",
      "MNLHRCPAGFCRRYTTESYILYALHRDNEGIAYTCAKLSPMVVGCILKDTG\n",
      "ME\n",
      "MTGGFLFCILSLDQITPIGGLEDPHHSRNFGTSMDWRPTGTLDAVASCASVLG\n",
      "MPCGLKLG\n",
      "MPMLWAILRNNNFPEIVVLYSIAGIFNRLLTSPKLQTTRHPGRDGPNNWPFQPVSFSIHPTTIGDSLAHVYAMPSLSL\n",
      "MGHSAQ\n",
      "MCQAITDGCRMYTKGYGLKRPVIWPISSGMPCGLKLG\n",
      "MGQITGRFSPYPLVYIRQPSVIAWHMYMLCPHYPCEERIICSFL\n",
      "MVSPFI\n",
      "MTRRFNKPGAVIWLVSA\n",
      "MLCPHYPCEERIICSFL\n",
      "MVVGCILKDTG\n",
      "MYLSASNPCWSRSF\n",
      "MLVPKFLEWCGSSRPPIGVI\n",
      "MPFYKWSLLCLLDEWRHHSDSARSYEPNLNGVAISCRSFTSSGLNPSPSDSLSL\n",
      "MHYNDRLSEGLGFNPEEVKDLQLMATPFRFGS\n",
      "MLWAILRNNNFPEIVVLYSIAGIFNRLLTSPKLQTTRHPGRDGPNNWPFQPVSFSIHPTTIGDSLAHVYAMPSLSL\n",
      "MPSLSL\n"
     ]
    }
   ],
   "source": [
    "import Bio.SeqIO\n",
    "# translating all possible open reading frames of a protein sequence from a\n",
    "#given string, including the complement strand.\n",
    "def codons(seq,frame): #requires a provided sequence and the frame of interest(frames being 1,2, or 3\n",
    "    n = len(seq) #gets length of the sequence\n",
    "    #print(n)\n",
    "    cdns = []\n",
    "    for i in range(frame - 1, n - 2, 3): #starts at (frame of interest) - 1, ends at (n-2) to prevent dicodons, and steps every 3.\n",
    "        cdns.append(seq[i:i+3])\n",
    "        #yield seq[i:i+3] #yields in specific and returns a generator object. Changed because it got rough after this\n",
    "    return cdns\n",
    "def get_lookup():\n",
    "    # Obtained from https://stackoverflow.com/questions/19521905/translation-dna-to-protein\n",
    "    return {\n",
    "    'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "    'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "    'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "    'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',\n",
    "    'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "    'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "    'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "    'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "    'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "    'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "    'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "    'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "    'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "    'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "    'TAC':'Y', 'TAT':'Y', 'TAA':'*', 'TAG':'*',\n",
    "    'TGC':'C', 'TGT':'C', 'TGA':'*', 'TGG':'W',\n",
    "    }\n",
    "def reverse_sequence(seq):\n",
    "    sequence = seq\n",
    "    rev_seq = ''\n",
    "    for nuc in seq:\n",
    "        if nuc == 'C':\n",
    "            rev_seq = rev_seq + 'G'\n",
    "        if nuc == 'G':\n",
    "            rev_seq = rev_seq + 'C'\n",
    "        if nuc == 'T':\n",
    "            rev_seq = rev_seq + 'A'\n",
    "        if nuc == 'A':\n",
    "            rev_seq = rev_seq + 'T'\n",
    "    rev_seq = rev_seq[::-1]  # inverts sequence to so that it's going in the right order\n",
    "    return rev_seq\n",
    "def orftranslate(seq):\n",
    "    aadict = get_lookup()\n",
    "    orflibrary = []  # list of translated orf\n",
    "    for frm in range(1, 4): #range() is exlusive on the end, so it needed to be 4\n",
    "        codonlist = codons(seq, frm) #populates codonlist with the codons of that frame\n",
    "        #print(codonlist)\n",
    "        for n in range(len(codonlist)):\n",
    "            if codonlist[n] == 'ATG':\n",
    "                templist = codonlist[n:]\n",
    "                #print(templist)\n",
    "                currentorf = []\n",
    "                for codon in templist:\n",
    "                    if codon == 'TAA' or codon == 'TGA' or codon == 'TAG':\n",
    "                        currentorf.append(aadict[codon])\n",
    "                        break\n",
    "                    else:\n",
    "                        currentorf.append(aadict[codon])\n",
    "                if ('*' in currentorf) == False: #needs to have a stop codon in the provided string to be valid\n",
    "                    break\n",
    "                else:\n",
    "\n",
    "                    aaframe = \"\".join(currentorf[:-1:])\n",
    "                    orflibrary.append(aaframe)\n",
    "            else:\n",
    "                continue\n",
    "    #for read in orflibrary:\n",
    "        #print(read)\n",
    "    return orflibrary\n",
    "for record in Bio.SeqIO.parse('rosalind_orf.txt', \"fasta\"):\n",
    "    fileseq = record.seq\n",
    "\n",
    "rev_fileseq = reverse_sequence(fileseq)\n",
    "final = orftranslate(fileseq)\n",
    "final.extend(orftranslate(rev_fileseq))\n",
    "for protein in set(final):\n",
    "    print(protein)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MRA\n",
      "MNGDTIQIRLVATSRI\n",
      "MTGCPKDWDLIPRR\n",
      "MLIPRLVLAPN\n",
      "MPRRILSPLHYRKLHIIRSSQG\n",
      "MYMLCPHYPCEERIICSFL\n",
      "MSEDG\n",
      "MATPFRFGS\n",
      "MYTKGYGLKRPVIWPISSGMPCGLKLG\n",
      "MKIHLYSFRNTKLQCTTMTGCPKDWDLIPRR\n",
      "MVSP\n",
      "MDWRPTGTLDAVASCASVLG\n",
      "MPAR\n",
      "M\n",
      "MNLHRCPAGFCRRYTTESYILYALHRDNEGIAYTCAKLSPMVVGCILKDTG\n",
      "ME\n",
      "MTGGFLFCILSLDQITPIGGLEDPHHSRNFGTSMDWRPTGTLDAVASCASVLG\n",
      "MPCGLKLG\n",
      "MPMLWAILRNNNFPEIVVLYSIAGIFNRLLTSPKLQTTRHPGRDGPNNWPFQPVSFSIHPTTIGDSLAHVYAMPSLSL\n",
      "MGHSAQ\n",
      "MCQAITDGCRMYTKGYGLKRPVIWPISSGMPCGLKLG\n",
      "MGQITGRFSPYPLVYIRQPSVIAWHMYMLCPHYPCEERIICSFL\n",
      "MVSPFI\n",
      "MTRRFNKPGAVIWLVSA\n",
      "MLCPHYPCEERIICSFL\n",
      "MVVGCILKDTG\n",
      "MYLSASNPCWSRSF\n",
      "MLVPKFLEWCGSSRPPIGVI\n",
      "MPFYKWSLLCLLDEWRHHSDSARSYEPNLNGVAISCRSFTSSGLNPSPSDSLSL\n",
      "MHYNDRLSEGLGFNPEEVKDLQLMATPFRFGS\n",
      "MLWAILRNNNFPEIVVLYSIAGIFNRLLTSPKLQTTRHPGRDGPNNWPFQPVSFSIHPTTIGDSLAHVYAMPSLSL\n",
      "MPSLSL\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "import re\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "f = \"rosalind_orf.txt\"\n",
    "\n",
    "with open(f) as file:\n",
    "    for record in SeqIO.parse(file, \"fasta\"):\n",
    "        seq = record.seq\n",
    "\n",
    "seq_2 = seq.reverse_complement()\n",
    "\n",
    "motive = re.compile(r\"(?=(ATG((?!(TGA)|(TAG)|(TAA))...)*(?=(TGA)|(TAG)|(TAA))))\")\n",
    "seqs1 = motive.finditer(str(seq))\n",
    "seqs_2 = motive.finditer(str(seq_2))\n",
    "seq1 = [match.group(1) for match in seqs1]\n",
    "seq_2 = [match.group(1) for match in seqs_2]\n",
    "seq = seq1 + seq_2\n",
    "\n",
    "orf = [str(Seq(s).translate()) for s in seq]\n",
    "\n",
    "orf = set(orf)\n",
    "\n",
    "[print(x) for x in orf]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SPLC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MSTFFLGPKSISTVIFRIQASRVTVAYGHSEAVGPFGARCWWLVVALERPRTPYGLNTRTVAQRFSNTQRHVSSFRAVFARDRKSKLKQSWGQWREDFIRQEMLTITIPVLHTRAGIPDSVFSHRVYPSHRIDLLPTEFTLVVEVKSDGYTKEDTDNDGAYMARVSLASI\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO                          \n",
    "from Bio.Seq import Seq                        \n",
    "lst = []                                 \n",
    "handle = open('rosalind_splc.txt', 'r')         \n",
    "for record in SeqIO.parse(handle, 'fasta'):    \n",
    "    seq = ''                              \n",
    "    for x in record.seq:                      \n",
    "        seq += x                        \n",
    "\n",
    "\n",
    "    lst.append(seq)                 \n",
    "handle.close()                                 \n",
    "DNA = lst[0]                        \n",
    "introns =lst[1:]                        \n",
    "\n",
    "for i in range(len(introns)):    \n",
    "\n",
    "    DNA= DNA.replace(introns[i], '')\n",
    "\n",
    "DNA = Seq(DNA)  \n",
    "                  \n",
    "print(DNA.translate(to_stop=True))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PMCH \n",
    "\n",
    "rectify"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from math import factorial as fac\n",
    "\n",
    "\n",
    "with open(\"rosalind_pmch.txt\") as inFile:\n",
    "\n",
    "\n",
    "    RNA = \"\".join(line.strip() for line in inFile.readlines()[1:])\n",
    "\n",
    "    \n",
    "with open(\"/Users/amelchoudary/Documents/University/YEAR 2/Computer Science  II/Assignments_CS_py/Assignment 5/answer_output\", \"w\") as outFile:\n",
    "\n",
    "    CG = RNA.count(\"C\")\n",
    "    AU = RNA.count(\"A\")\n",
    "\n",
    "    matching = fac(CG) * fac(AU)\n",
    "    print(f\"{matching}\", file=outFile)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PPER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "\n",
    "def func(list, size_perm, target):\n",
    "     #Generate all permutations\n",
    "     perms = itertools.permutations(list, size_perm)\n",
    "     \n",
    "     #iterate of all permutations and print only if sum == target\n",
    "     for item in perms:\n",
    "         if sum(item) == target:\n",
    "             print(item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "N = [0.5, 0.8, 1, 1.5, 2, 2.5, 3, 5, 7.5, 10, 12.5, 15, 17.5, 20, 25, 30]\n",
    "func(N, 21, 20)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "84b7f5baf7709d8b0040a7b044644fcec20d91e73b471742cc0a411029aab504"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
